src/Qowaiv.DomainModel/Internal/AppendOnlyCollection.cs

```diff
@@ -0,0 +1,167 @@
+
+    [DebuggerBrowsable(DebuggerBrowsableState.Never)]
+    private readonly object[] Buffer;
+
+    /// <summary>Initializes a new instance of the <see cref="AppendOnlyCollection"/> struct.</summary>
+    private AppendOnlyCollection(int count, object[] buffer)
+    {
+        Count = count;
+        Buffer = buffer;
+    }
+
+    /// <inheritdoc />
+    public int Count { get; }
+
+    /// <summary>Creates a new <see cref="AppendOnlyCollection"/> with the added item(s).</summary>
+    /// <param name="item">
+    /// The item(s) to add.
+    /// </param>
+    /// <remarks>
+    /// Null, and null items are ignored.
+    /// </remarks>
+    [Pure]
+    public AppendOnlyCollection Add(object? item)
+    {
+        if (item is null)
+        {
+            return this;
+        }
+        if (item is IEnumerable enumerable && item is not string)
+        {
+            return AddRange(enumerable.GetEnumerator());
+        }
+        else
+        {
+            return AddSingle(item);
+        }
+    }
+
+    /// <summary>Returns a specified range of contiguous elements from the collection.</summary>
+    [Pure]
+    public Enumerator Take(int count) => new(Buffer, Math.Min(count, Count));
+
+    /// <summary>
+    /// Bypasses a specified number of elements in the collection and then returns the remaining elements.
+    /// </summary>
+    [Pure]
+    public Enumerator Skip(int count) => new(Buffer, count, Count);
+
+    /// <summary>Adds a single item to the collection.</summary>
+    /// <remarks>
+    /// If the parent collection has not been extended yet, the array buffer is
+    /// shared with the new collection (if it still fits), otherwise the buffer
+    /// is copied first.
+    /// </remarks>
+    [Pure]
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private AppendOnlyCollection AddSingle(object element)
+    {
+        var append = Ensure();
+
+        if (append.Buffer != Buffer)
+        {
+            return append.Append(element);
+        }
+        else
+        {
+            lock (Buffer)
+            {
+                if (Buffer[Count] is null)
+                {
+                    return Append(element);
+                }
+            }
+            return Copy(Buffer.Length).Append(element);
+        }
+    }
+
+    /// <summary>Adds multiple elements to the collection.</summary>
+    /// <remarks>
+    /// Adding the first item is done like adding a single item: by doing so, a
+    /// lock is only required for that first item, because whatever the outcome,
+    /// the remaining items can always be added without checking the lock.
+    /// </remarks>
+    [Pure]
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private AppendOnlyCollection AddRange(IEnumerator iterator)
+    {
+        var append = this;
+
+        while (iterator.MoveNext())
+        {
+            if (iterator.Current is { } element)
+            {
+                append = append.AddSingle(element);
```

Comment:
Not locking the entire while loop could cause odd behaviour if multiple threads try to insert at the same time. It could lead to a different set adding an element in between an `AddRange`.

Reply:
It should not: once we added one item, we can be certain that other threats will conclude that this will never be  their buffer anymore.